src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,443 @@
+ * <p>
+ * TODO: finish this javadoc
+ */
+
+public final class KDF {
+    private static final Debug debug = Debug.getInstance("jca", "KeyDerivation");
+
+    private static final Debug pdebug = Debug.getInstance("provider", "Provider");
+    private static final boolean skipDebug = Debug.isOn("engine=") && !Debug.isOn("keyderive");
+
+    // The provider
+    private Provider provider;
+
+    // The provider implementation (delegate)
+    private KDFSpi spi;
+
+    // The name of the MAC algorithm.
+    private final String algorithm;
+
+    // Additional KDF configuration parameters
+    private final AlgorithmParameterSpec algorithmParameterSpec;
+
+    // Has this object been initialized?
+    private boolean initialized = false;
+
+    // next service to try in provider selection
+    // null once provider is selected
+    private Service firstService;
+
+    // remaining services to try in provider selection
+    // null once provider is selected
+    private Iterator<Service> serviceIterator;
+
+    private final Object lock;
+
+    /**
+     * Instantiates a KeyDerivation object.
+     *
+     * @param keyDerivSpi
+     *     the delegate
+     * @param provider
+     *     the provider
+     * @param algorithm
+     *     the algorithm
+     * @param algParameterSpec
+     *     the algorithm parameters
+     */
+    protected KDF(KDFSpi keyDerivSpi, Provider provider, String algorithm,
+                  AlgorithmParameterSpec algParameterSpec) {
+        this.spi = keyDerivSpi;
+        this.provider = provider;
+        this.algorithm = algorithm;
+        this.algorithmParameterSpec = algParameterSpec;
+        this.lock = null;
+    }
+
+    /**
+     * Returns the algorithm name of this {@code KeyDerivation} object.
+     *
+     * <p>This is the same name that was specified in one of the
+     * {@code getInstance} calls that created this {@code KeyDerivation} object.
+     *
+     * @return the algorithm name of this {@code KeyDerivation} object.
+     */
+    public final String getAlgorithm() {
+        return this.algorithm;
+    }
+
+    /**
+     * Returns the provider of this {@code KeyDerivation} object.
+     *
+     * @return the provider of this {@code KeyDerivation} object.
+     */
+    private final Provider getProvider() {
+        chooseFirstProvider();
+        return this.provider;
+    }
+
+    private String getProviderName() {
+        return (provider == null) ? "(no provider)" : provider.getName();
+    }
+
+    /**
+     * Creates an instance of the {@code KDF} object.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if no {@code Provider} supports a {@code KDFSpi} implementation for the specified
+     *     algorithm.
+     */
+    public static KDF getInstance(String algorithm) throws NoSuchAlgorithmException {
```

Comment:
All the `getInstance` methods should throw `NullPointerException` if `algorithm` is `null`.